BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Authentication/ReflectionCredentialEnvelope.cs

```diff
@@ -0,0 +1,203 @@
+
+            /// This creates a wrapper for the following method:
+            /// <code>public abstract Azure.Core.AccessToken GetToken (Azure.Core.TokenRequestContext requestContext, System.Threading.CancellationToken cancellationToken).</code>
+            /// (https://docs.microsoft.com/dotnet/api/azure.core.tokencredential.gettoken).
+            private static Delegate BuildDelegateGetToken()
+            {
+                Type typeTokenCredential = Type.GetType("Azure.Core.TokenCredential, Azure.Core");
+                Type typeTokenRequestContext = Type.GetType("Azure.Core.TokenRequestContext, Azure.Core");
+                Type typeCancellationToken = typeof(CancellationToken);
+
+                var parameterExpression_tokenCredential = Expression.Parameter(type: typeTokenCredential, name: "parameterExpression_TokenCredential");
+                var parameterExpression_requestContext = Expression.Parameter(type: typeTokenRequestContext, name: "parameterExpression_RequestContext");
+                var parameterExpression_cancellationToken = Expression.Parameter(type: typeCancellationToken, name: "parameterExpression_CancellationToken");
+
+                var exprGetToken = Expression.Call(
+                    instance: parameterExpression_tokenCredential,
+                    method: typeTokenCredential.GetMethod(name: "GetToken", types: new Type[] { typeTokenRequestContext, typeCancellationToken }),
+                    arg0: parameterExpression_requestContext,
+                    arg1: parameterExpression_cancellationToken);
+
+                var exprTokenProperty = Expression.Property(
+                    expression: exprGetToken,
+                    propertyName: "Token");
+
+                return Expression.Lambda(
+                    body: exprTokenProperty,
+                    parameters: new ParameterExpression[]
+                    {
+                        parameterExpression_tokenCredential,
+                        parameterExpression_requestContext,
+                        parameterExpression_cancellationToken,
+                    }).Compile();
+            }
+
+            /// <summary>
+            /// This is a wrapper for the following method:
+            /// <code>public abstract System.Threading.Tasks.ValueTask&lt;Azure.Core.AccessToken> GetTokenAsync (Azure.Core.TokenRequestContext requestContext, System.Threading.CancellationToken cancellationToken);</code>
+            /// (https://docs.microsoft.com/dotnet/api/azure.core.tokencredential.gettokenasync).
+            /// </summary>
+            /// <returns>
+            /// The Expression Tree library cannot handle async methods.
+            /// As a workaround, this method returns two Delegates. 
+            /// First;
+            /// The first Delegate is a wrapper around GetTokenAsync which returns a ValueTask of AccessToken.
+            /// Then calls ValueTask.GetTask to convert that to a Task which is a known type for older frameworks.
```

Comment:
Is this step (converting ValueTask to Task) necessary? Could you use `ValueTask<AccessToken>` directly?
i.e. Have the first delegate/expression return the `ValueTask<AccessToken>` and pull out the `.Result.Token` of that in the 2nd delegate.

Reply:
I've currently got no way to `await` the `ValueTask` because that type isn't known to these older frameworks.
Also the Expression Tree api doesn't provide support for async methods.
I'm very open to suggestions here! :)

Reply:
Ah, I see. The await is the problem. You need to return something "awaitable" (has a compiler-visible GetAwaiter() method). Yup. Task is probably the best way.